tachyon/zk/plonk/examples/example_circuit_tests.cc

```diff
@@ -0,0 +1,536 @@
+#include <type_traits>
+#include <typeinfo>
+#include <utility>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+#include "tachyon/base/array_to_vector.h"
+#include "tachyon/math/elliptic_curves/bn/bn254/bn254.h"
+#include "tachyon/zk/base/commitments/gwc_extension.h"
+#include "tachyon/zk/base/commitments/shplonk_extension.h"
+#include "tachyon/zk/lookup/halo2/scheme.h"
+#include "tachyon/zk/plonk/examples/circuit_test.h"
+#include "tachyon/zk/plonk/examples/simple_circuit.h"
+#include "tachyon/zk/plonk/examples/simple_circuit_test_data.h"
+#include "tachyon/zk/plonk/halo2/pinned_constraint_system.h"
+#include "tachyon/zk/plonk/halo2/pinned_verifying_key.h"
+#include "tachyon/zk/plonk/keys/proving_key.h"
+#include "tachyon/zk/plonk/layout/floor_planner/simple_floor_planner.h"
+#include "tachyon/zk/plonk/layout/floor_planner/v1/v1_floor_planner.h"
+
+namespace tachyon::zk::plonk::halo2 {
+
+namespace {
+
+using SHPlonk =
+    SHPlonkExtension<math::bn254::BN254Curve, kMaxDegree, kMaxExtendedDegree,
+                     math::bn254::G1AffinePoint>;
+
+using GWC = GWCExtension<math::bn254::BN254Curve, kMaxDegree,
+                         kMaxExtendedDegree, math::bn254::G1AffinePoint>;
+
+// Field is currently the same across SHPlonk and GWC
+using F = typename SHPlonk::Field;
+
+// Currently Halo2 lookup's |Poly|, |Evals|, and |Commitment| are the same as
+// SHPlonks's/GWC's
+using Halo2LS =
+    lookup::halo2::Scheme<typename SHPlonk::Poly, typename SHPlonk::Evals,
+                          typename SHPlonk::Commitment>;
+
+template <typename _Circuit, typename _PCS, typename _LS>
+struct TestingArguments {
+  using Circuit = _Circuit;
+  using PCS = _PCS;
+  using LS = _LS;
+};
+
+template <typename Circuit, typename PCS, typename LS>
+struct ExampleTestDataSelector {
+  using F = typename PCS::Field;
+  // SimpleTestData -> SimpleFloorPlanner
+  // SimpleTestData -> V1FloorPlanner
+
+  // clang-format off
+  using type =
+      std::conditional_t<
+        std::is_same_v<Circuit, SimpleCircuit<F, SimpleFloorPlanner>>,
+          SimpleTestData<SimpleCircuit<F, SimpleFloorPlanner>, PCS, LS>,
+          SimpleTestData<SimpleCircuit<F, V1FloorPlanner>, PCS, LS>>;
```

Comment:
It would be nice to have a Trait class to not depend on a concrete class `SimpleCircuit` to decide `TestData` type.

Reply:
Also, it would be nice to change the name to `Type`.